Development/nmos/streamcompatibility_behaviour.cpp

```diff
@@ -0,0 +1,306 @@
+            {
+                return std::vector<nmos::id>::value_type{ resource.id };
+            }));
+        }
+
+        nmos::sender_state validate_sender_resources(const web::json::value& transport_file, const web::json::value& flow, const web::json::value& source, const web::json::array& constraint_sets)
+        {
+            nmos::sender_state sender_state;
+
+            if (!web::json::empty(constraint_sets))
+            {
+                bool constrained = true;
+
+                auto source_found = std::find_if(constraint_sets.begin(), constraint_sets.end(), [&](const web::json::value& constraint_set) { return match_source_parameters_constraint_set(source, constraint_set); });
+                auto flow_found = std::find_if(constraint_sets.begin(), constraint_sets.end(), [&](const web::json::value& constraint_set) { return match_flow_parameters_constraint_set(flow, constraint_set); });
+
+                constrained = constraint_sets.end() != source_found && constraint_sets.end() != flow_found;
+
+                if (!transport_file.is_null() && !transport_file.as_object().empty())
+                {
+                    utility::string_t sdp_data = get_sdp_data(transport_file);
+                    const auto session_description = sdp::parse_session_description(utility::us2s(sdp_data));
+                    auto sdp_params = nmos::parse_session_description(session_description).first;
+
+                    const auto format_params = nmos::details::get_format_parameters(sdp_params);
+                    const auto sdp_found = std::find_if(constraint_sets.begin(), constraint_sets.end(), [&](const web::json::value& constraint_set) { return nmos::details::match_sdp_parameters_constraint_set(nmos::details::format_constraints, sdp_params, format_params, constraint_set); });
+
+                    constrained = constrained && constraint_sets.end() != sdp_found;
+                }
+
+                sender_state = constrained ? nmos::sender_states::constrained : nmos::sender_states::active_constraints_violation;
+            }
+            else
+            {
+                sender_state = nmos::sender_states::unconstrained;
+            }
+
+            return sender_state;
+        }
+
+        nmos::receiver_state validate_receiver_resources(const web::json::value& transport_file, const web::json::value& receiver)
+        {
+            nmos::receiver_state receiver_state;
+
+            if (!transport_file.is_null() && !transport_file.as_object().empty())
+            {
+                utility::string_t sdp_data = get_sdp_data(transport_file);
+                const auto session_description = sdp::parse_session_description(utility::us2s(sdp_data));
+                auto sdp_params = nmos::parse_session_description(session_description).first;
+
+                receiver_state = nmos::receiver_states::compliant_stream;
+
+                try
+                {
+                    validate_sdp_parameters(receiver, sdp_params);
```

Comment:
This needs to be configurable by the user as well, I think?

Reply:
Yes. Made `validate_receiver_resources()` a configurable callback.

Reply:
Similarly to https://github.com/sony/nmos-cpp/pull/271/files#r1028471346, I wonder if we want a `make_streamcompatibility_receiver_validator` function that takes the SDP parser and validator to prevent the user having to write this whole function.

Reply:
Added `make_streamcompatibility_receiver_validator` that takes `nmos::transport_file_parser` and it has some restrictions:
* this function is expected to parse and validate the transport file at once. Not sure if it can be split into two arguments because I can't imagine a validator that is unaware of the internal details of the transport file format.
* functions of this type take `gate` but `parse_rtp_transport_file` (the only function of this type?) doesn't actually use it.

Reply:
> * functions of this type take `gate` but `parse_rtp_transport_file` (the only function of this type?) doesn't actually use it.

Hmm, yeah, I guess that seemed like a good idea at the time... https://github.com/sony/nmos-cpp/commit/5bc3962e42aa5af01410fef5a1b5fec499b28043 😄
We don't like making incompatible changes if we can help it, but if the `gate` parameter is causing problems, we should revisit.

> * this function is expected to parse and validate the transport file at once. 

OK, I'll review to understand what you mean!